src/responder/common/cache_req/cache_req_search.c

```diff
@@ -84,6 +84,149 @@ static void cache_req_search_ncache_add(struct cache_req *cr)
     return;
 }
 
+static errno_t
+cache_req_search_get_name_from_msg(TALLOC_CTX *mem_ctx,
+                                   struct ldb_message *msg,
+                                   struct sss_domain_info *domain,
+                                   bool override_space,
+                                   char **_name)
+{
+    TALLOC_CTX *tmp_ctx;
+    const char *name;
+    char *cased_name;
+    errno_t ret;
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    name = sss_get_name_from_msg(domain, msg);
+    if (name == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "sss_get_name_from_msg() returned NULL, which should never "
+              "happen in this scenario!\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    cased_name = sss_get_cased_name(tmp_ctx, name, domain->case_preserve);
+    if (cased_name == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+                "sss_get_cased_name failed, skipping\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    cased_name = sss_replace_space(tmp_ctx, cased_name, override_space);
+    if (name == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "sss_replace_space failed\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    *_name = talloc_steal(mem_ctx, cased_name);
+    ret = EOK;
+
+done:
+    talloc_free(tmp_ctx);
+    return ret;
+}
+
+static errno_t cache_req_search_ncache_filter(TALLOC_CTX *mem_ctx,
+                                              struct cache_req *cr,
+                                              struct ldb_result *result,
+                                              struct ldb_result **_result)
+{
+    TALLOC_CTX *tmp_ctx;
+    struct ldb_result *filtered_result;
+    struct ldb_message **msgs;
+    size_t msg_count;
+    char *name = NULL;
+    errno_t ret;
+
+    if (cr->plugin->ncache_filter_fn == NULL) {
+        CACHE_REQ_DEBUG(SSSDBG_TRACE_FUNC, cr,
+                        "This request type does not support filtering negative cache\n");
+        *_result = result;
+        return EOK;
+    }
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    CACHE_REQ_DEBUG(SSSDBG_TRACE_FUNC, cr,
+                    "Filtering out results from negative cache\n");
+
+    filtered_result = talloc_zero(mem_ctx, struct ldb_result);
+    if (filtered_result == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    msgs = talloc_zero(tmp_ctx, struct ldb_message *);
```

Comment:
Because this function will filter out only users and groups from the configuration file, there won't be many objects filtered out. But there will be many allocations with enumeration. It would be better to allocate enough space to hold all the messages at once to safe some allocations. I.e.
```c
msgs = talloc_zero_array(tmp_ctx, struct ldb_message *, result->count)
```
And remove the reallocs.